tests/code_update.py

```diff
@@ -51,6 +51,180 @@ def test_verify_quotes(network, args):
+        else "libjs_generic"
+    )
+
+    primary, others = network.find_nodes()
+    prev_nodes = [primary, *others]
+    new_node = network.create_node("local://localhost")
+    network.join_node(new_node, replacement_package, args, timeout=3, env=dict())
+    network.trust_node(new_node, args)
+    primary, others = network.find_nodes()
+    nodes = [primary, *others]
+    assert len(nodes) == len(prev_nodes) + 1, "New node not added to network"
+
+    with primary.client() as client:
+        r = client.get("/gov/security_policy")
+        policies = sorted(r.body.json()["policies"], key=lambda x: x["digest"])
+
+    LOG.info(f"Checking governance table security policies")
+    expected = [
+        {
+            "digest": DEFAULT_SNP_SECURITY_POLICY_DIGEST,
+            "raw": DEFAULT_SNP_SECURITY_POLICY,
+        }
+    ]
+    expected.sort(key=lambda x: x["digest"])
+    assert policies == expected, [(a, b) for a, b in zip(policies, expected)]
+
+    LOG.info(f"Checking node joined with a blank security policy")
+    # TODO: Do this
+
+@reqs.description("Node where raw security policy doesn't match digest fails to join")
+def test_add_node_with_mismatched_security_policy_digest(network, args):
+
+    if not IS_SNP:
+        LOG.warning("Skipping test_add_node_with_mismatched_security_policy_digest with non SNP platform")
+        return network
+
+    replacement_package = (
+        "samples/apps/logging/liblogging"
+        if args.package == "libjs_generic"
+        else "libjs_generic"
+    )
+
+    primary, others = network.find_nodes()
+    prev_nodes = [primary, *others]
+    new_node = network.create_node("local://localhost")
+    try:
+        network.join_node(new_node, replacement_package, args, timeout=3, env={"SECURITY_POLICY": "eyJhbGxvd19hbGwiOmZhbHNlLCJjb250YWluZXJzIjp7Imxlbmd0aCI6MCwiZWxlbWVudHMiOm51bGx9fQ=="})
+        network.trust_node(new_node, args)
+    except Exception as err:
+        ...
+
+    primary, others = network.find_nodes()
+    nodes = [primary, *others]
+    assert len(nodes) == len(prev_nodes), "Node joining unexpectedly succeeded"
+
+@reqs.description("Node with bad security policy digest fails to join")
+def test_add_node_with_bad_security_policy_digest(network, args):
+
+    if not IS_SNP:
+        LOG.warning("Skipping test_add_node_with_bad_security_policy_digest with non SNP platform")
+        return network
+
+    replacement_package = (
+        "samples/apps/logging/liblogging"
+        if args.package == "libjs_generic"
+        else "libjs_generic"
+    )
+
+    primary, others = network.find_nodes()
+    prev_nodes = [primary, *others]
+
+    LOG.info(f"Removing security policy set by node 0 so that a new joiner is seen as an unmatching policy")
+    network.consortium.retire_security_policy(primary, DEFAULT_SNP_SECURITY_POLICY_DIGEST)
+
+    new_node = network.create_node("local://localhost")
+    network.join_node(new_node, replacement_package, args, timeout=3)
```

Comment:
I would expect this call to timeout in this case?

Reply:
Unfortunately in it's current state it's joining successfully, I suspect something to do with the table not being up to date everywhere when we retire a policy?

Reply:
Yes this is unexpected. AIUI, the call to `join_node` should time out here?

Reply:
Yeah I think fixing the governance stuff while changing digest to a SHA256 fixed this so now it does time out as expected